src/disk_cache.rs

```diff
@@ -0,0 +1,956 @@
+            }
+        }
+
+        Ok(part_numbers)
+    }
+
+    // if the disk is full, we'll get an error here.
+    // TODO: this error can be ignored in the caller side and print a warning.
+    async fn save_part(&self, part_number: usize, buf: &[u8]) -> object_store::Result<()> {
+        let part_file_path = self.make_part_path(part_number, false);
+
+        // if the part already exists, do not save again.
+        if Some(true) == fs::try_exists(&part_file_path).await.ok() {
+            return Ok(());
+        }
+
+        // ensure the parent folder exists
+        if let Some(part_file_folder) = part_file_path.parent() {
+            fs::create_dir_all(part_file_folder)
+                .await
+                .map_err(wrap_io_err)?;
+        }
+
+        // save the file content to tmp. if the disk is full, we'll get an error here.
+        let tmp_part_file_path =
+            part_file_path.with_extension(format!("tmp{}", self.make_rand_suffix()));
+        let mut file = OpenOptions::new()
+            .write(true)
+            .create(true)
+            .truncate(true)
+            .open(&tmp_part_file_path)
+            .await
+            .map_err(wrap_io_err)?;
+        file.write_all(buf).await.map_err(wrap_io_err)?;
+
+        // rename the file
+        fs::rename(tmp_part_file_path, part_file_path)
+            .await
+            .map_err(wrap_io_err)
+            .ok();
+        Ok(())
+    }
+
+    async fn save_meta(&self, meta: &ObjectMeta) -> object_store::Result<()> {
+        let meta_file_path = self.make_meta_path();
+
+        // if the meta file exists and not corrupted, do nothing
+        match self.read_meta().await {
+            Ok(Some(_)) => return Ok(()),
+            Ok(None) => {}
+            Err(_) => {
+                // TODO: add a warning
+            }
+        }
+
+        // ensure the parent folder exists
+        if let Some(part_file_folder) = meta_file_path.parent() {
+            fs::create_dir_all(part_file_folder)
+                .await
+                .map_err(wrap_io_err)?;
+        }
+
+        let meta = DiskCacheMeta {
+            location: meta.location.to_string(),
+            last_modified: meta.last_modified.to_rfc3339(),
+            size: meta.size,
+            e_tag: meta.e_tag.clone(),
+            version: meta.version.clone(),
+        };
+        let buf = serde_json::to_vec(&meta).map_err(wrap_io_err)?;
+
+        let tmp_meta_file_path =
+            meta_file_path.with_extension(format!("tmp{}", self.make_rand_suffix()));
+        let mut file = OpenOptions::new()
+            .write(true)
+            .create(true)
+            .truncate(true)
+            .open(&tmp_meta_file_path)
+            .await
+            .map_err(wrap_io_err)?;
+        file.write_all(&buf).await.map_err(wrap_io_err)?;
+
+        // rename the tmp file to meta_file_path
+        if fs::try_exists(&meta_file_path).await.unwrap_or(false) {
```

Comment:
Why bother removing the file if `rename` overwrites it anyway?

Reply:
i asked gpt and it said that `fs::rename` might raise an error on some systems if the target file already existed, it suggested that removing before `rename` might have a better compatibility across different systems (but i dunno it is hallucination or not):

> Prompt:
> what will happen if the target file exists while calling fs::rename in rust?
>
> GPT-4o:
> If the target file exists when calling fs::rename in Rust, it will be replaced by the source file. This operation is atomic, meaning it's done in a way that ensures the integrity of the operation, but the behavior might vary across different platforms.
> 
> On Unix-like systems, fs::rename will overwrite the target. On Windows, if the target exists, the operation will fail, as Windows does not allow renaming a file to an existing file without explicitly removing the target first.
> 
> To handle this, you might need to check if the target exists and handle it accordingly, such as removing the target file before renaming.

Reply:
Hmmm... that sounds like a hallucination to me. According to the [Rust docs](https://doc.rust-lang.org/std/fs/fn.rename.html):

> This function currently corresponds to the rename function on Unix and the MoveFileEx function with the MOVEFILE_REPLACE_EXISTING flag on Windows.

There is some nuance around renaming across mount points, but that shouldn't affect us.

Reply:
haha, thank you for the infomation, i'll fix it